tephra-api/src/main/java/co/cask/tephra/Transaction.java

```diff
@@ -48,35 +63,69 @@
+   *                      For new transactions, this will be the same as {@code writePointer}.  For checkpointed
+   *                      transactions, this will be the most recent write pointer issued.
+   * @param invalids list of invalid transactions to exclude while reading
+   * @param inProgress list of in-progress transactions to exclude while reading
+   * @param firstShortInProgress earliest in-progress short transaction
+   * @param type transaction type
+   * @param checkpointPointers the list of writer pointers added from checkpoints on the transaction
+   */
+  public Transaction(long readPointer, long writePointer, long currentWritePointer, long[] invalids, long[] inProgress,
+                     long firstShortInProgress, TransactionType type, long[] checkpointPointers) {
     this.readPointer = readPointer;
     this.writePointer = writePointer;
+    this.currentWritePointer = currentWritePointer;
     this.invalids = invalids;
     this.inProgress = inProgress;
     this.firstShortInProgress = firstShortInProgress;
     this.type = type;
+    this.checkpointWritePointers = checkpointPointers;
   }
 
   /**
-   * Creates a new short transaction.
-   * @param readPointer read pointer for transaction
-   * @param writePointer write pointer for transaction
-   * @param invalids list of invalid transactions to exclude while reading
-   * @param inProgress list of in-progress transactions to exclude while reading
-   * @param firstShortInProgress earliest in-progress short transaction
+   * Creates a new transaction for a checkpoint operation, copying all members from the original transaction,
+   * with the updated checkpoint write pointers.
+   *
+   * @param toCopy the original transaction containing the state to copy
+   * @param currentWritePointer the new write pointer to use for the transaction
+   * @param checkpointPointers the list of write pointers added from checkpoints on the transaction
    */
-  public Transaction(long readPointer, long writePointer, long[] invalids, long[] inProgress,
-                     long firstShortInProgress) {
-    this(readPointer, writePointer, invalids, inProgress, firstShortInProgress, TransactionType.SHORT);
+  public Transaction(Transaction toCopy, long currentWritePointer, long[] checkpointPointers) {
+    this(toCopy.getReadPointer(), toCopy.getWritePointer(), currentWritePointer, toCopy.getInvalids(),
+        toCopy.getInProgress(), toCopy.getFirstShortInProgress(), toCopy.getType(), checkpointPointers);
   }
 
   public long getReadPointer() {
     return readPointer;
   }
 
+  /**
+   * Returns the initial write pointer assigned to the transaction.  This will remain the same for the life of the
+   * transaction, and uniquely identifies it with the transaction service.  This value should be provided
+   * to identify the transaction when calling any transaction lifecycle methods on the transaction service.
+   */
```

Comment:
This is now not the write version any more, but it serves to identify the transaction. Therefore I would suggest to name it the transaction ID and call this method getTransactionId.